OPHAKTKEH-111

pkoivisto opened the pull request.

pkoivisto committed:
Remove leftover components/translators/..

pkoivisto committed:
Add internationalizations

pkoivisto committed:
Interfaces

pkoivisto committed:
Use "content-type: application/json" with axios

pkoivisto committed:
Add enums

pkoivisto committed:
ContactRequest redux logic & sagas

pkoivisto committed:
UI navigation state within redux

pkoivisto committed:
Further redux changes

pkoivisto committed:
Update existing components

pkoivisto committed:
New styles

pkoivisto committed:
Implement ContactRequestForm

pkoivisto committed:
Align data models between frontend and backend

pkoivisto committed:
Remove comment, add missing email address

pkoivisto committed:
Add basic input validation to contact details fields

pkoivisto committed:
Improve validation: keep next button disabled if any required field is invalid, but only show error marker on field if it has been changed

pkoivisto committed:
Fix rendering translators: filter by chosen language pair as well

pkoivisto committed:
Each step must consider on its own if the next button should be disabled

p3j4m commented on src/main/reactjs/public/i18n/fi-FI.json:
```diff
@@ -97,6 +100,41 @@
+          "4": "Valmis"
+        },
+        "chosenTranslatorsForLanguagePair": "Valitut kääntäjät kieliparista",
+        "firstName": "Etunimi",
+        "lastName": "Sukunimi",
+        "email": "Sähköpostiosoite",
+        "phoneNumber": "Puhelinnumero",
+        "contactInfo": "Yhteystiedot",
+        "message": "Viesti",
+        "writeMessageHereLabel": "Kirjoita viestisi tähän",
```
voisi siirtää formin in alle eli forum { .... }

pkoivisto replied:
siirsin `formLabels` alle

p3j4m commented on src/main/reactjs/public/i18n/fi-FI.json:
```diff
@@ -97,6 +100,41 @@
+          "submit": "Lähetä"
+        },
+        "steps": {
+          "0": "Vahvista kääntäjät",
+          "1": "Täytä yhteystiedot",
+          "2": "Kirjoita viesti",
+          "3": "Esikatsele ja lähetä",
+          "4": "Valmis"
```
Pitäisikö käyttää kirjaimia/sanoja numeroiden sijaan?

pkoivisto replied:
Muutettu.

p3j4m commented on src/main/reactjs/src/components/tables/Table.tsx:
```diff
@@ -45,34 +47,39 @@ export function PaginatedTable<T>({
+        <div className="rows">
+          <div className="columns">
+            <div className="grow">{showNumOfSelected()}</div>
+            {callToAction}
```
Tarvitseeko tämän olla osaa Table komponenttia? Se lisää monimutkaisuutta Table komponenttiin eikä lisää välttämättä arvoa. Jos on tehty designing takia voidaan katsoa designia uudelleen. Tämä monimutkaistuu vielä kun listataan valitut kääntäjät seuraavasti:
![image](https://user-images.githubusercontent.com/84386398/146504299-032364f4-a024-4b5b-bff4-88d7efb41168.png)

pkoivisto replied:
Joo, eihän tuon välttämättä tarvitse. Siirsin tän pelkästään `TranslatorListing.tsx` puolelle.

p3j4m commented on src/main/reactjs/src/components/translator/TranslatorListing.tsx:
```diff
@@ -9,13 +15,15 @@ import { ProgressIndicator } from 'components/elements/ProgressIndicator';
   removeSelectedTranslator,
 } from 'redux/actions/translatorDetails';
 import { publicTranslatorsSelector } from 'redux/selectors/translatorDetails';
+import { UiStates } from 'enums/app';
```
Olisi hyvä kirjoittaa isoilla kirjaimilla kuten Airbnb:n best practicissa sanotaan --> acronyms and initialisms should always be all capitalized,

p3j4m commented on src/main/reactjs/src/enums/app.ts:
```diff
@@ -5,3 +5,8 @@ export enum AppConstants {
   HomePage = '/',
 }
+
+export enum UiStates {
```
UiStates --> UIStates

pkoivisto replied:
Muutan.

p3j4m commented on src/main/reactjs/src/interfaces/uiState.ts:
```diff
@@ -0,0 +1,5 @@
+import { UiStates } from 'enums/app';
+
+export interface UiState {
```
UiState --> UIState

pkoivisto replied:
Muutan.

p3j4m commented on src/main/reactjs/src/pages/HomePage.tsx:
```diff
@@ -11,23 +11,52 @@ import {
+  );
+};
+
+export const HomePage: FC = () => {
```
HomePAge  --> PublicTranslatorHomePage ?

Samanlainen homepage tulee myös virkailija käyttöliittymään.

pkoivisto replied:
Nuo translator + home page yhdessä ei jotenkin istu mun korvaan hyvin. :)
Miltä kuulostaisi esim. PublicLandingPage (vs. ClerkLandingPage) tai PublicHomePage tai PublicTranslatorList(Page)?

pkoivisto replied:
Se on nyt `PublicTranslatorsHomePage`

p3j4m replied:
Olet oikeassa PublicHomePage kuulostaa paremmalta :)

p3j4m commented on src/main/reactjs/src/pages/HomePage.tsx:
```diff
@@ -11,23 +11,52 @@ import {
+import { ContactRequestForm } from 'components/contactRequest/ContactRequestForm';
 
-export const HomePage: FC = () => {
+export const TranslatorsGrid = () => {
```
Tämän komponentin voisi siirtää components --> translator alle

pkoivisto replied:
Muutettu.

p3j4m commented on src/main/reactjs/src/components/contactRequest/ContactRequestForm.tsx:
```diff
@@ -0,0 +1,562 @@
+import {
```
Kannattaa pilkkoa ContactRequestForm.tsx pienempiin osiin. contactRequest kansion alla voisi olla muita tähän liittyviä tiedostoja.

pkoivisto replied:
Muutettu. Tähän jäi nyt lähinnä stepperin koordinointi ja vähän layoutia.

p3j4m commented on src/main/reactjs/src/redux/actions/contactRequest.ts:
```diff
@@ -0,0 +1,16 @@
+  CONTACT_REQUEST_SET,
+} from 'redux/actionTypes/contactRequest';
+
+export const setContactRequest = (request: ContactRequest) => {
+  return { type: CONTACT_REQUEST_SET, request };
+};
+
+export const sendContactRequest = (request: ContactRequest) => {
+  return { type: CONTACT_REQUEST_SEND, request };
+};
```
returnit pois sen sijaan voi käyttää ()

p3j4m commented on src/main/reactjs/src/redux/reducers/navigation.ts:
```diff
@@ -0,0 +1,17 @@
+
+const defaultState = { state: UiStates.PublicTranslatorListing };
+
+export const uiStateReducer: Reducer<UiState, Action> = (
```
Tässäkin käyttäisin UI

pkoivisto replied:
Muutan.

p3j4m commented on src/main/reactjs/src/styles/components/dialog/_dialog.scss:
```diff
@@ -0,0 +1,11 @@
+.dialog__success {
+  .MuiDialog-paper {
+    border-top: 1vh $color-green-600 solid;
+  }
+}
+
+.dialog__error {
+  .MuiDialog-paper {
+    border-top: 1vh $color-red-500 solid;
+  }
+}
```
Kannattaisi laittaa nämä kaikki .dialog in alle. CSS:ssä material UI:n classien käyttö aiheuttaa ongelmia. Kannattaisi välttää niitä ja sen sijaan luoda omia classeja

pkoivisto replied:
Siirsin .dialog alle ja sinne tuli vielä kolmantena luokkana `dialog__neutral`. En saanut tyylejä tarttumaan käyttämättä tuota .MuiDialog-paper -luokkaa, joten jätin vielä tuollaisiksi.

p3j4m replied:
Käyttäisin esim:

```
@mixin top-border($color) {
  border-top: 1rem $color solid;
}

.dialog {
  &__success[role='dialog'] {
    @include top-border($color-some-some);
  }

  &__error[role='dialog'] {
    @include top-border($color-some-some);
  }

  &__info[role='dialog'] {
    @include top-border($color-some-some);
  }
}

```

dialog__info on parempi ehkä kuin dialog__neutral?

pkoivisto replied:
Nice, toi role-attribuutti vaikuttaa hyvältä. `dialog__info` myös parempi kuin `dialog__neutral`.

pkoivisto replied:
Ei ihan vielä tuollaisenaan toiminut, piti vielä muuttaa `[role='dialog']` noiden luokkien descendantiksi. Poistin myös border-radiukset, kun wireframeissakaan niitä ei ollut.

p3j4m commented on src/main/reactjs/src/components/contactRequest/ContactRequestForm.tsx:
```diff
@@ -0,0 +1,562 @@
+const incrementStep = (step: number) => step + 1;
+
+const ChosenTranslatorsHeading = () => {
+  const fromLang = useAppSelector(
+    (state) => state.translatorDetails.filters.fromLang
+  );
+  const toLang = useAppSelector(
+    (state) => state.translatorDetails.filters.toLang
+  );
```
Käytetään globaaleja selectoreja eli 

{filters } = useAppSelector(publicTranslatorsSelector) yllä olevian sijaan.

Saman periaatteen voisi käyttää kaikissa muissa useAppSelector kutsuissakin.

pkoivisto replied:
Muutan.

p3j4m commented on src/main/reactjs/src/components/contactRequest/ContactRequestForm.tsx:
```diff
@@ -0,0 +1,562 @@
+    keyPrefix: 'akt.component.contactRequestForm',
+  });
+  const contactInfo = useAppSelector(
+    (state) => state.contactRequest.request
```
Käytetään globaaleja selectoreja

pkoivisto replied:
Muutan.

p3j4m commented on src/main/reactjs/src/components/contactRequest/ContactRequestForm.tsx:
```diff
@@ -0,0 +1,562 @@
+  contactDetailsField: keyof ContactDetails;
+  isValidCallback: (field: keyof ContactDetails, error: boolean) => void;
+}) => {
+  const contactDetails = useAppSelector(
```
Katso rivi 121. Parempi käyttää tässäkin valmiita selectoreja.

pkoivisto replied:
Muutan.

pkoivisto committed:
Review feedback: capitalization

pkoivisto committed:
Review feedback: prefer global selectors

pkoivisto committed:
Review feedback: prefer implicit returns

pkoivisto committed:
Review feedback: extract components to separate files from ContactRequestForm.tsx

pkoivisto committed:
Simplify logic for deciding whether to show a dialog or spinner.

Use the APIResponseStatus values directly instead of separate booleans for each state of the enum.
Allow the dialog wrapper components to directly control their own open states.

pkoivisto committed:
Extract connected dialog wrappers to their own file

pkoivisto committed:
Extract step definitions from ContactRequestForm.tsx -> Steps.tsx

pkoivisto committed:
Review feedback: extract TranslatorsGrid, rename HomePage to PublicTranslatorsHomePage

pkoivisto committed:
Review feedback: adjust i18n label structure

pkoivisto committed:
Implement cancel confirmation dialog

pkoivisto committed:
Destructure rest of props used within the StyledDialog component

pkoivisto committed:
Review feedback: make stepper step label i18n keys more descriptive

pkoivisto committed:
Review feedback: don't embed extra features to PaginatedTable

pkoivisto committed:
Review feedback: use attribute selector instead of material-UI classnames

pkoivisto committed:
Use separate mixin for removing border radius

p3j4m commented on src/main/reactjs/public/i18n/fi-FI.json:
```diff
@@ -22,6 +22,9 @@
         },
         "selectedItems": "valittu"
       },
+      "callToAction": {
```
Käytetäänkö tätä enää?

pkoivisto replied:
Eipä käytetä. Poistan.

p3j4m commented on src/main/reactjs/src/components/contactRequest/ConnectedDialog.tsx:
```diff
@@ -0,0 +1,115 @@
+  };
+
+  return (
+    <NeutralDialog
```
Tämän pitäisi olla varmaan myös InfoDialog?

pkoivisto replied:
Jep, muutan.

p3j4m commented on src/main/reactjs/src/components/contactRequest/ConnectedDialog.tsx:
```diff
@@ -0,0 +1,115 @@
+export const CancelRequestDialog = ({
+  open,
+  onClose,
+}: {
+  open: boolean;
+  onClose: () => void;
+}) => {
+  const { t } = useAppTranslation({
+    keyPrefix: 'akt.component.contactRequestForm.cancelRequestDialog',
+  });
+  const dispatch = useAppDispatch();
+  const cleanUp = () => {
+    dispatch(resetContactRequest);
+    dispatch(displayUIState(UIStates.PublicTranslatorListing));
+    onClose();
+  };
+
+  return (
+    <NeutralDialog
+      title={t('title')}
+      content={<Text>{t('description')}</Text>}
+      actions={
+        <>
+          <Button variant="outlined" color="secondary" onClick={onClose}>
+            {t('back')}
+          </Button>
+          <Button
+            variant="contained"
+            color="secondary"
+            onClick={() => cleanUp()}
+          >
+            {t('yes')}
+          </Button>
+        </>
+      }
+      open={open}
+      onClose={onClose}
+    />
+  );
+};
```
Hyvältä vaikuttaa. Voidaan miettiä kuitenkin noiden yhdistämistä taskissa OPHAKTKEH-93

p3j4m commented on src/main/reactjs/src/components/contactRequest/ContactRequestForm.tsx:
```diff
@@ -0,0 +1,171 @@
+      setContactRequest({
+        languagePair: { from, to },
+        translatorIds: selectedTranslators,
+        email: '',
+        firstName: '',
+        lastName: '',
+        message: '',
+        phoneNumber: '',
```
Voisiko asettaa näitä contactRequest reducerissa alussa ns. defaultStatena? Silloin näitä ei tarvitse asettaa enää tässä.

pkoivisto replied:
Joo, varmaankin. Ehkä pitää tietomallia muokata jotenkin, koska tosiaan mm. tuo kielipari on nyt ContactRequestin pakollinen kenttä, ja sitähän ei tiedetä ennen kuin käyttäjä on tehnyt valinnan. :) Muut kentät voi varmaan olla tyhjää stringiä tai tyhjää arrayta defaultina.

pkoivisto committed:
Review feedback: remove obsolete i18n entry

pkoivisto committed:
Review feedback: NeutralDialog -> InfoDialog

p3j4m commented on src/main/reactjs/src/components/contactRequest/ContactRequestForm.tsx:
```diff
@@ -0,0 +1,171 @@
+          <Box className="contact-request-form__form-container-box">
+            <Box className="contact-request-form__inner-content-box">
+              <ContactRequestStepper step={step} />
+              {step == 0 && (
+                <VerifyTranslatorsStep disableNext={disableNextCb} />
+              )}
+              {step == 1 && (
+                <FillContactDetailsStep disableNext={disableNextCb} />
+              )}
+              {step == 2 && <WriteMessageStep disableNext={disableNextCb} />}
+              {step == 3 && <PreviewAndSendStep />}
```
Tekisin vielä stepeistä enumin esim. ContactRequestFormStep tms. 

step == ContactRequestFormStep.First 

Sen voisi ottaa käyttöön myös Rivillä 64.

p3j4m commented on src/main/reactjs/src/components/contactRequest/Steps.tsx:
```diff
@@ -0,0 +1,271 @@
+  });
+  return (
+    <H3>
+      {t('contactRequestForm.chosenTranslatorsForLanguagePair')}{' '}
+      {t('publicTranslatorFilters.languages.' + fromLang)} -{' '}
+      {t('publicTranslatorFilters.languages.' + toLang)}{' '}
```
```
      {t('publicTranslatorFilters.languages.' + fromLang)} -{' '}  voisi kirjoittaa esim.
  {`${t(`publicTranslatorFilters.languages.${fromLang}`)} - `} 

```


 Toisaalta en ole varma onko toista helpompi lukea :) Molemmat on ok.

p3j4m commented on src/main/reactjs/src/components/contactRequest/ValidatedContactDetailsField.tsx:
```diff
@@ -0,0 +1,45 @@
+      label={t(contactDetailsField)}
+      value={contactDetails[contactDetailsField]}
+      onChange={(e) => {
+        setFieldStateChanged(true);
+        const updated = Object.assign({}, contactDetails);
+        updated[contactDetailsField] = e.target.value;
+        dispatch(setContactRequest(updated));
```
Voisi ehkä tehdä oman funktion noiden suorittamista varten.

p3j4m commented on src/main/reactjs/src/pages/PublicTranslatorsHomePage.tsx:
```diff
@@ -0,0 +1,36 @@
+import { FC, useEffect } from 'react';
+import { Box, Grid } from '@mui/material';
+
+import { useAppDispatch, useAppSelector } from 'configs/redux';
+import { UIStateSelector } from 'redux/selectors/navigation';
+import { UIStates } from 'enums/app';
+import { ContactRequestForm } from 'components/contactRequest/ContactRequestForm';
+import { PublicTranslatorsGrid } from 'components/translator/PublicTranslatorsList';
+import { loadTranslatorDetails } from 'redux/actions/translatorDetails';
+
+export const HomePage: FC = () => {
+  const dispatch = useAppDispatch();
+
+  useEffect(() => {
+    dispatch(loadTranslatorDetails);
+  }, [dispatch]);
+
+  const { state: currentUIState } = useAppSelector(UIStateSelector);
+
+  return (
+    <Box className="homepage">
+      <Grid
+        container
+        rowSpacing={4}
+        direction="column"
+        className="homepage__grid-container"
+      >
+        {currentUIState == UIStates.ContactRequest ? (
+          <ContactRequestForm />
+        ) : (
+          <PublicTranslatorsGrid />
+        )}
+      </Grid>
+    </Box>
+  );
+};
```
Sun ehdotus oli parempi. Tiedoston nimi voisi olla PublicHomePage.

p3j4m commented on src/main/reactjs/src/styles/components/dialog/_dialog.scss:
```diff
@@ -0,0 +1,30 @@
+@mixin top-border($color) {
+  border-top: 1rem $color solid;
+}
+
+@mixin no-border-radius {
+  border-radius: 0px;
+}
```
Pitäisikö yhdistää nämä yhdeksi mixiniksi

pkoivisto replied:
Mietin samaa, mutta sitten oli vaikea keksiä järkevää yhteistä nimeä niille.

p3j4m replied:
dialog-border voisi olla esim yksi vaihtoehto :)

pkoivisto replied:
Hyvä ehdotus. Muutan.

p3j4m commented on src/main/reactjs/src/components/contactRequest/Steps.tsx:
```diff
@@ -0,0 +1,271 @@
+            isValidCallback={memoizedFieldErrorCallback}
+            required
+          />
+          <ValidatedContactDetailsField
```
type attribuutin voisi välittää myös. Saadaan HTML email tarkistus päälle.

pkoivisto replied:
Pelkkä typen välitys TextFieldille ei taida riittää validaatioon: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/email#pattern_validation mukaan pitäisi tämän lisäksi määritellä itse vaaditut rajoitteet emailille, esim. `pattern=...`.

p3j4m replied:
Joo, vaikuttaisi siltä, ett' vaikka se lisää sen input elementtiin niin ei kuitenkaan näytä HTML erroreja. Palataan vaikka tähän myöhemmin eri taskin yhteydessä.

p3j4m reviewed (approved):
Näyttää hyvältä! Pieniä korjauksia voisi tehdä vielä ja samalla voisi varmistaa että buildi menee läpi.

pkoivisto committed:
Review feedback: combine mixins & rename

pkoivisto committed:
Review feedback: rename to PublicHomePage

pkoivisto committed:
Review feedback: pass field type to underlying TextField component for email address input

pkoivisto committed:
Review feedback: extract onChange handler to a function

pkoivisto committed:
Refactor conditional rendering logic for step contents and status notifications into their separate components

pkoivisto committed:
Pass explicit minStep to stepper controls

pkoivisto committed:
Fix build: add new file to git

pkoivisto committed:
Run prettier
